### PR TITLE
SJRK-405: Implement server side logic for signup/login/logout and session support

### PR DESCRIPTION
**Pull request by jobara**
https://issues.fluidproject.org/browse/SJRK-405

### CODE REVIEW

File: src/server/dataSource.js
Comment (BlueSlug) [inline src/server/dataSource.js]: Could you file a Jira for this and link to it from this comment, please?

File: src/server/requestHandlers.js
Comment (BlueSlug) [inline src/server/requestHandlers.js]:
```suggestion
// Kettle request handler for getting a single story (Edit)
```

File: src/server/requestHandlers.js
Comment (BlueSlug) [inline src/server/requestHandlers.js]: No URL rebuilding is happening any more as of `0.4.0`, I forgot to remove that part of the comment for `sjrk.storyTelling.server.handleGetStory`. This part of the sentence should be removed for that function as well as this one here

File: src/server/requestHandlers.js
Comment (BlueSlug) [inline src/server/requestHandlers.js]: This comment style (along with its partner at the bottom) doesn't match the rest of the project. I see that it's being used to delineate which grades are handlers for stories & files and which are for session management, and I wonder if it might not make more sense to split these into their own file, e.g. `requestHandlers-session.js` and `requestHandlers-story.js`, or some other more meaningful/practical naming convention. What do you think?

File: src/server/serverSetup.js
Comment (BlueSlug) [inline src/server/serverSetup.js]:
```suggestion
                        target: "{that expressUserUtils}.options.dataSourceConfig.host"
```

File: src/server/requestHandlers.js
Comment (BlueSlug) [inline src/server/requestHandlers.js]: In other parts of the codebase, I've put `{Component}` in cases where the arg is an Infusion component. For consistency, I'd suggest either setting this to that or setting all the other ones to the fully-qualified grade name. The latter might be out of scope for this Jira, in which case we should file a new one to look into what we should be doing and then implementing whatever solution we come up with.

File: src/server/serverSetup.js
Comment (BlueSlug) [inline src/server/serverSetup.js]: Is 10 iterations enough? I've seen 100, 1,000 or even 10,000 iterations suggested in order to defend against brute-force attacks, especially so with faster hashing algorithms. That said, if we're heading towards cloud hosting, too high a number could make hosting very expensive...

File: src/server/serverSetup.js
Comment (BlueSlug) [inline src/server/serverSetup.js:356]: If this `TODO` is not intended to be tackled in SJRK-405, please file a Jira to do that work and add a link to that Jira here

File: src/server/serverSetup.js
Comment (BlueSlug) [inline src/server/serverSetup.js]: Please add a JSDoc comment for this function

File: tests/server/dataSource-tests.js
Comment (BlueSlug) [inline tests/server/dataSource-tests.js]: It may be helpful to name these test cases after their contents or purpose

File: tests/server/dataSource-tests.js
Comment (BlueSlug) [inline tests/server/dataSource-tests.js]: A similar filtering out of `_rev` is done in `sjrk.storyTelling.server.testServerWithStorageDefs.verifyStoryPersistence`. Perhaps that piece of code and this function could be combined in a shared function in `serverTestUtils.js`.

File: tests/server/dataSource-tests.js
Comment (BlueSlug) [inline tests/server/dataSource-tests.js]: Would the expected values be better defined outside of this function and passed in?

**Review by BlueSlug — COMMENTED** [review]
Looking pretty good!

File: src/server/dataSource.js
Comment (jobara) [reply src/server/dataSource.js]: https://issues.fluidproject.org/browse/SJRK-443

File: src/server/serverSetup.js
Comment (jobara) [reply src/server/serverSetup.js]: This is the default from the underlying library actually. We could make it some other value as needed.

File: src/server/serverSetup.js
Comment (jobara) [reply src/server/serverSetup.js:356]: https://issues.fluidproject.org/browse/SJRK-444

File: src/server/requestHandlers.js
Comment (jobara) [reply src/server/requestHandlers.js]: I've switched it to `{Component}` for consistency. We should review this in the future. In Infusion the practice has moved to use the component grade; however, a description is still required by the JSDoc validation.

File: tests/server/dataSource-tests.js
Comment (BlueSlug) [inline tests/server/dataSource-tests.js]: Perhaps a distributeOptions could be used to inject this host configuration into each test dataSource. This is fine either way, though :)

File: tests/server/dataSource-tests.js
Comment (BlueSlug) [inline tests/server/dataSource-tests.js:221]: Is this meant to read "ddoc" or "doc"? If it's the former, it may be more clear if it says "design doc"

Comment (jobara) [reply tests/server/dataSource-tests.js:221]: that's an error straight from the database, it isn't something I defined.

Comment (BlueSlug) [reply tests/server/dataSource-tests.js:221]: Curious! But okay, let's just leave it as-is, in that case

File: src/server/db/authors-dbConfiguration.js
Comment (cindyli) [inline src/server/db/authors-dbConfiguration.js]: Just curious, is "authors" a separated database that only contains the "user" document type? What's the reason to decide on using a separate database rather than making use of the existing "stories" database with a new "user" document type?

With the current logic, using 2 databases works the same way. But it feels strange using multiple databases to support an app, which is unnecessary.

File: src/server/db/singleNodeConfiguration.js
Comment (cindyli) [inline src/server/db/singleNodeConfiguration.js]: What is this "users" database? It doesn't seem to be used anywhere.

Comment (jobara) [reply src/server/db/singleNodeConfiguration.js]: I believe it's couchDB's user database for accounts to itself.

Comment (jobara) [reply src/server/db/singleNodeConfiguration.js]: oh actually this one might be a mistake, might have been leftovers from trying things out early on.

File: src/server/db/authors-dbConfiguration.js
Comment (jobara) [reply src/server/db/authors-dbConfiguration.js]: That the user accounts are managed through fluid-express-users library, but I think it could actually be in the same database as stories. I'll look into that.

File: src/server/storyRequestHandlers.js
Comment (cindyli) [inline src/server/storyRequestHandlers.js:237]: Could you add a code comment here and at line 226 that this line is to update an existing story while line 226 is to create a new story.

**Comment by cindyli** [comment]
It would be helpful to add server API documentation when you have time. Thanks.

File: src/server/db/authors-dbConfiguration.js
Comment (BlueSlug) [reply src/server/db/authors-dbConfiguration.js]: I looked into this when I first saw @jobara's choice. Turns out using separate databases is actually preferable, on account of the way index building works in CouchDB, though the performance gains to us will be minimal. Here's a very interesting thread on the topic, I recommend reading the answers other than the one the OP chose: https://stackoverflow.com/questions/10503089/couchdb-multiple-databases

Comment (jobara) [reply src/server/db/authors-dbConfiguration.js]: @BlueSlug that's a very interesting point. Unfortunately for me I already combined them. @cindyli and @BlueSlug let me know if you think we should split them back out.

Comment (cindyli) [reply src/server/db/authors-dbConfiguration.js]: It's great to know separating data into multiple databases helps with the CouchDB performance. Thanks for the good thread, @BlueSlug. After reading the thread, I'm fine with either.

If @jobara has already combined it, I would prefer leave as combined. As @BlueSlug said, the performance gain for SJRK will be minimal. Moreover, using one database simplifies the code and is easier for backup and maintenance.

**Comment by jobara** [comment]
@cindyli this PR is ready for review.

File: src/server/db/story-dbConfiguration.js
Comment (cindyli) [inline src/server/db/story-dbConfiguration.js:154]: The condition of checking `doc.value.published` shows a more meaning name for this view might be "publishedStoriesById`.

Comment (jobara) [reply src/server/db/story-dbConfiguration.js:154]: Makes sense. I've renamed it.

File: src/server/serverSetup.js
Comment (cindyli) [inline src/server/serverSetup.js]: What is this account for? `secrets.json.example` does not provide an example of this account value.

This affects line 69-72 too which distributes this value.

File: src/server/serverSetup.js
Comment (cindyli) [inline src/server/serverSetup.js:287]: This pull request does not contain the tests and documentation for these new endpoints. Is it planned to be addressed in a separate ticket?

File: src/server/db/story-dbConfiguration.js
Comment (cindyli) [inline src/server/db/story-dbConfiguration.js]: I know nothing about the client side logic. As the server logic in this pull request always returns the email address as a login success signal, I wonder if a user could login via either the username or email address, or just email address. In the case of the latter, it might be better to remove this line in order to reduce the index volume and prevent a potential security issue.

File: src/server/serverSetup.js
Comment (jobara) [reply src/server/serverSetup.js]: This provides a means of configuring [`fluid.express.user.utils`](https://github.com/fluid-project/fluid-express-user/blob/e826290beb4fdb416a4fedf602803ac1bcef4f37/src/js/server/utils.js#L28). Not sure how to show that in the `secrets.json.example`. Any suggestions?

File: src/server/serverSetup.js
Comment (jobara) [reply src/server/serverSetup.js:287]: The authentication related handlers are included in [testSeverWithStorage.js](https://github.com/jobara/sjrk-story-telling/blob/SJRK-405/tests/server/testServerWithStorage.js) as they are required for saving/editing stories.

File: src/server/db/story-dbConfiguration.js
Comment (jobara) [reply src/server/db/story-dbConfiguration.js]: There are no explicit usernames anymore, only an e-mail address to identify authors. Thanks for spotting this leftover bit. I've removed it, and have also renamed the name of the function. There are some other references to byUsernameOrEmail but these are used internally by fluid-express-user, so I have left them as is.

File: src/server/serverSetup.js
Comment (jobara) [reply src/server/serverSetup.js:287]: They are also tested in [/testServerAuthentication.js](https://github.com/jobara/sjrk-story-telling/blob/SJRK-405/tests/server/testServerAuthentication.js)

File: src/server/serverSetup.js
Comment (cindyli) [reply src/server/serverSetup.js]: Can you point me to the doc or code that show how this `account` is used for configuring `fluid.express.user.utils`? I didn't find its usage in the given link. Just want to get an idea about its value type and the logic behind it. Thanks.

Comment (jobara) [reply src/server/serverSetup.js]: @cindyli the link above is to the components defaults. We add fluid.express.user.utils as a [subcomponent](https://github.com/jobara/sjrk-story-telling/blob/89fe0ec255b7fef24aa12c107a7f3b198ac257df/src/server/serverSetup.js#L76-L102) to the server. These defaults can be set/overridden by passing in configuration to the secrets.json file under the `account` property. These are relayed via the lines you mentioned above. The component options for [account](https://github.com/jobara/sjrk-story-telling/blob/89fe0ec255b7fef24aa12c107a7f3b198ac257df/src/server/serverSetup.js#L55) and the [options distributions](https://github.com/jobara/sjrk-story-telling/blob/89fe0ec255b7fef24aa12c107a7f3b198ac257df/src/server/serverSetup.js#L69-L72) to send this to the instance of fluid.express.user.utils.

**Comment by jobara** [comment]
@cindyli I'm trying to write up some docs on the server endpoints as well. Working through how to format them, but I have an idea now.

**Comment by jobara** [comment]
@cindyli I've added the server endpoints documentation. Ready for more review.

**Comment by cindyli** [comment]
The server endpoints doc looks great. Thanks, @jobara.

File: src/server/serverSetup.js
Comment (cindyli) [reply src/server/serverSetup.js]: I think now I understand better of what you said about "provides a means of configuring fluid.express.user.utils". A couple of suggestions:

1. Please consider to use a more meaningful variable name such as `expressUserUtilsConfig`;
2. I'm not sure if it's a good idea to use `secrets.json` to distribute component options. These options only make sense to SJRK developers while `secrets.json` gives a feeling of being configured by website admins who unlikely know about infusion and what component options are;
3. If you decided to use `secrets.json` for this config, please add documentation, for example in README where `secrets.json` is explained. (If `secrets.json` is a json5 file, comments for each value in that file would be easier to access.)

Comment (jobara) [reply src/server/serverSetup.js]: @cindyli I've added some documentation about configuration, add some examples to the secrets.json.example file, and renamed some things. I think mostly you can just use the kettle config to configure expressUserUtils, but there are some parts that we may want to express in the secrets file, so I've illustrated those in the example.

File: src/server/serverSetup.js
Comment (cindyli) [inline src/server/serverSetup.js]: The differentiation of the kettle server config and account credential config in various config files makes a lot of sense. Thanks, @jobara.

With the renaming at line 55, the source reference of `account` should also be renamed to `authorCredentialConfig`. It would be nice to have tests for the proper handling/distribution of secrets.json.

Comment (jobara) [reply src/server/serverSetup.js]: Thanks for catching the leftover bit. I've addressed that and added some test for the distributions of the secrets.

**Comment by cindyli** [comment]
This pull request looks great to me. 

@BlueSlug, can you help to do a full manual test? Once it passes the test, feel free to merge. Thanks.

File: docs/CONFIGURATION.md
Comment (BlueSlug) [inline docs/CONFIGURATION.md:67]: The section of `README.md` titled "Running the site" mentions what the contents of `secrets.json` need to be in order to run the site. I suggest either updating that section to reflect the contents here, or linking directly to this file

**Review by BlueSlug — APPROVED** [review]
The tests are all passing without issue, now :)

**Comment by jobara** [comment]
I believe @BlueSlug gave his [approval](https://matrix.to/#/!mDigWuPWilYGiEcOCA:matrix.org/$8lcQ77Vz-nE2jIlbd6AQEq1lv3l7XjkG163ta1EeYUg?via=matrix.org) in the channel. I'll self merge this PR.

**Comment by jobara** [comment]
Merged at 3845afd05c22ba96283c4558581b90e69ea47319

**Pull request merged**
